src/com/ka/spreadsheet/diff/UnifiedDiffSpreadSheetDiffCallback.java

```diff
@@ -0,0 +1,168 @@
+//        ...
+// Each cell data line is always present, even if the data is empty (thus just "-" or "+).
+public class UnifiedDiffSpreadSheetDiffCallback implements SpreadSheetDiffCallback {
+
+  private final String lineSeparator = System.getProperty("line.separator");
+  private String file1;
+  private String file2;
+  private CellPos previousCell = null;
+  private DiffCell prevDiffCell = new DiffCell("", -2, -2, null, null);
+  private List<DiffCell> currentCellBlock = new ArrayList<DiffCell>();
+
+  @Override
+  public void init(String file1, String file2) {
+    this.file1 = file1;
+    this.file2 = file2;
+  }
+
+  @Override
+  public void finish() {
+    printAndEmptyCellBlock();
+  }
+
+  @Override
+  public void reportWorkbooksDiffer(boolean differ) {
+  }
+
+  @Override
+  public void reportMacroOnlyIn(boolean inFirstSpreadSheet) {
+    System.out.println("Unified diff format does not support macros, however WB" + (inFirstSpreadSheet ? "1" : "2") + " contains at least one macro that is not in the other workbook.");
+  }
+
+  @Override
+  public void reportExtraCell(boolean inFirstSpreadSheet, CellPos c) {
+    assert previousCell == null || c.compareCellPositions(previousCell) >= 0 :
+      "Cell-ordering contract violated.  Previous=" + previousCell.getCellPosition()
+      + ", current=" + c.getCellPosition();
+    previousCell = c;
+    accumulateAndMaybePrint(new DiffCell(
+      c.getSheetName(),
+      c.getRowIndex(),
+      c.getColumnIndex(),
+      (inFirstSpreadSheet ? c.getCellValue() : null),
+      (!inFirstSpreadSheet ? c.getCellValue() : null)
+    ));
+  }
+
+  @Override
+  public void reportDiffCell(CellPos c1, CellPos c2) {
+    assert (c1.getRowIndex() == c2.getRowIndex())
+      && (c1.getColumnIndex() == c2.getColumnIndex()) : "Cells are not at the same position. Cell 1="
+      + c1.getCellPosition() + ", cell 2=" + c2.getCellPosition();
+    assert previousCell == null || c1.compareCellPositions(previousCell) >= 0 :
+      "Cell-ordering contract violated.  Previous=" + previousCell.getCellPosition()
+      + ", current=" + c1.getCellPosition();
+    previousCell = c1;
+    accumulateAndMaybePrint(new DiffCell(
+      c1.getSheetName(),
+      c1.getRowIndex(),
+      c1.getColumnIndex(),
+      c1.getCellValue(),
+      c2.getCellValue()
+    ));
+  }
+
+  private void accumulateAndMaybePrint(DiffCell diffCell) {
+    if (!isSameSheet(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+      System.out.println("--- " + file1 + "!" + diffCell.sheetName);
+      System.out.println("+++ " + file2 + "!" + diffCell.sheetName);
+    }
+    if (!isSameRow(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+    }
+    if (!isSameCellBlock(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+    }
+    currentCellBlock.add(diffCell);
+    prevDiffCell = diffCell;
+  }
+
+  private boolean isSameSheet(DiffCell c1, DiffCell c2) {
+    assert c1 != null && c2 != null;
```

Comment:
these are never going to be null, you could inline these functions

Reply:
Yeah, I might have gone overboard with the `asserts` :-)  I refactored these methods out because I wanted the names to explain what they were doing, even though they were each being used in only one place.  I do that sometimes for code readability - I'm a retired developer and development manager, and I've had to read my own code 20 and 30 years later sometimes, and been confused :-)  It's up to you, I'll follow your lead on your project.

Reply:
I'll remove the assertions in the next PR, and leave the private methods for readability.